### PR TITLE
Multi runlocal

### PR DESCRIPTION
**Pull request by maasg**
This PR implements support for the new multi-project model in the _sandbox_ environment (aka `runLocal`).
To achieve classpath isolation, each sub-project of a multi-project application is executed in a separate JVM. They communicate through a common (local) Kafka instance that is managed by the plugin.

Note that this PR separates `LocalRunner` into a new project to avoid classpath pollution from the transitive dependencies that `LocalRunner` requires. In particular, `EmbeddedKafka`

### CODE REVIEW

File: core/build.sbt [hunk 1]
```diff
@@ -370,6 +366,33 @@ lazy val runner =
...
 
+lazy val localRunner =
+  cloudflowModule("cloudflow-localrunner")
+    .enablePlugins(BuildInfoPlugin, ScalafmtPlugin)
+    .dependsOn(streamlets, blueprint)
+    .settings(
+      scalafmtOnCompile := true,
+      libraryDependencies ++= Vector(
+            Ficus,
+            Slf4jLog4jBridge
```
Comment (RayRoestenburg) [inline core/build.sbt:377]: Have you tried to run this with a pure akka app?

File: core/cloudflow-spark/src/main/scala/cloudflow/spark/SparkStreamlet.scala [hunk 2]
```diff
@@ -95,16 +93,24 @@ trait SparkStreamlet extends Streamlet[SparkStreamletContext] with Serializable
...
-      val sysCheck: Cancellable = system.scheduler.schedule(1.minute, 1.minutes) {
-        val totalMem = Runtime.getRuntime.totalMemory() / 1024.0 / 1024.0
-        // TODO temporary logging total mem in case a recent issue resurfaces in a long running spark process.
-        log.info(s"jvm-total-mem: $totalMem Mb")
+      val sysCheck: Cancellable = system.scheduler.scheduleWithFixedDelay(1.minute, 1.minutes) {
+        new Runnable {
+          override def run(): Unit = {
+            val totalMem = Runtime.getRuntime.totalMemory() / 1024.0 / 1024.0
+            // TODO temporary logging total mem in case a recent issue resurfaces in a long running spark process.
+            log.info(s"jvm-total-mem: $totalMem Mb")
```
Comment (RayRoestenburg) [inline core/cloudflow-spark/src/main/scala/cloudflow/spark/SparkStreamlet.scala:111]: This needs to be removed again

File: core/sbt-cloudflow/src/main/scala/cloudflow/sbt/CloudflowBasePlugin.scala [hunk 3]
```diff
@@ -56,7 +56,8 @@ object CloudflowBasePlugin extends AutoPlugin {
           // this artifact needs to have `%` and not `%%` as we build the runner jar
           // without version information. This is required for Flink runtime as a fixed name
           // jar needs to be uploaded to a specific location for Flink operator to pick up
-          "com.lightbend.cloudflow" % "cloudflow-runner" % BuildInfo.version
+          "com.lightbend.cloudflow" % "cloudflow-runner"       % BuildInfo.version,
+          "com.lightbend.cloudflow" %% "cloudflow-localrunner" % BuildInfo.version
```
Comment (RayRoestenburg) [inline core/sbt-cloudflow/src/main/scala/cloudflow/sbt/CloudflowBasePlugin.scala:60]: Does this mean that localrunner and all of its dependencies land up in the image that is built?

File: core/sbt-cloudflow/src/main/scala/cloudflow/sbt/CloudflowKeys.scala [hunk 4]
```diff
@@ -75,7 +75,10 @@ trait CloudflowTaskKeys {
     taskKey[Map[String, Config]]("Streamlets found in sub projects by scanning the application classpath.")
 
   private[sbt] val cloudflowStreamletDescriptors = taskKey[Map[String, Config]]("Streamlets found by scanning the application classpath.")
-  private[sbt] val cloudflowApplicationClasspath = taskKey[Array[URL]]("classpath of the user project.")
+  private[sbt] val cloudflowStreamletDescriptorsByProject =
```
Comment (RayRoestenburg) [inline core/sbt-cloudflow/src/main/scala/cloudflow/sbt/CloudflowKeys.scala:78]: This looks the same as `allStreamletDescriptorsByProject`?

File: core/sbt-cloudflow/src/main/scala/cloudflow/sbt/CloudflowLocalRunnerPlugin.scala [hunk 5]
```diff
@@ -95,27 +155,118 @@ object CloudflowLocalRunnerPlugin extends AutoPlugin {
...
+                               appDescriptor: ApplicationDescriptor,
+                               appDescriptorFile: Path,
+                               outputFile: File,
+                               logConfig: Path,
+                               localConfMsg: String)
+
+  def setupKafka(port: Int, topics: Seq[String])(implicit log: Logger) = {
+    implicit val kafkaConfig = EmbeddedKafkaConfig(kafkaPort = port)
+    EmbeddedKafka.start()
+    println("topics in this app:" + topics.mkString(", "))
```
Comment (RayRoestenburg) [inline core/sbt-cloudflow/src/main/scala/cloudflow/sbt/CloudflowLocalRunnerPlugin.scala:174]: Is this debugging leftover maybe?

File: core/sbt-cloudflow/src/main/scala/cloudflow/sbt/CloudflowLocalRunnerPlugin.scala [hunk 6]
```diff
@@ -170,7 +322,7 @@ object CloudflowLocalRunnerPlugin extends AutoPlugin {
       .withOutputStrategy(OutputStrategy.LoggedOutput(logger))
       .withConnectInput(false)
       .withRunJVMOptions(Vector(s"-Dlog4j.configuration=file:///${log4JConfigFile.toFile.getAbsolutePath}"))
-    val classpathStr = classpath.map(url ⇒ new File(url.toURI)).mkString(separator)
+    val classpathStr = classpath.collect { case url if !url.toString.contains("logback") ⇒ new File(url.toURI) }.mkString(separator)
```
Comment (RayRoestenburg) [inline core/sbt-cloudflow/src/main/scala/cloudflow/sbt/CloudflowLocalRunnerPlugin.scala:325]: Assuming you then also add log4j to the classpath somewhere?

File: core/sbt-cloudflow/src/main/scala/cloudflow/sbt/CloudflowBasePlugin.scala (hunk 3 above)
Comment (maasg) [reply core/sbt-cloudflow/src/main/scala/cloudflow/sbt/CloudflowBasePlugin.scala:60]: Yes, but...
```
lazy val localRunner =
  cloudflowModule("cloudflow-localrunner")
    .enablePlugins(BuildInfoPlugin, ScalafmtPlugin)
    .dependsOn(streamlets, blueprint)
    .settings(
      scalafmtOnCompile := true
    )
```
I removed all dependencies from the build and, in particular, moved the log dependencies to the runtime.

File: core/sbt-cloudflow/src/main/scala/cloudflow/sbt/CloudflowKeys.scala (hunk 4 above)
Comment (maasg) [reply core/sbt-cloudflow/src/main/scala/cloudflow/sbt/CloudflowKeys.scala:78]: one is needed to populate the other.

File: core/sbt-cloudflow/src/main/scala/cloudflow/sbt/CloudflowLocalRunnerPlugin.scala (hunk 5 above)
Comment (maasg) [reply core/sbt-cloudflow/src/main/scala/cloudflow/sbt/CloudflowLocalRunnerPlugin.scala:174]: :grin:

File: integration-tests/swiss-knife/app/src/main/blueprint/blueprint.conf [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    spark-process = swissknife.spark.SparkCounter
+    spark-output = swissknife.spark.SparkOutput
+    flink-process = swissknife.flink.FlinkCounter
+    akka-process = swissknife.akka.AkkaTransformation
+  }
+
+  topics {
+    source-data-pipe {
+      producers = [ingress.out]
+      consumers = [ egress.in, akka-process.in ,  spark-process.in , flink-process.in, spark-output.in]
```
Comment (RayRoestenburg) [inline integration-tests/swiss-knife/app/src/main/blueprint/blueprint.conf:15]:
```suggestion
      consumers = [egress.in, akka-process.in, spark-process.in, flink-process.in, spark-output.in]
```

File: integration-tests/swiss-knife/app/src/main/blueprint/blueprint.conf [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+    }
+
+    process-pipe {
+      producers = [ spark-process.out, akka-process.out, flink-process.out ]
```
Comment (RayRoestenburg) [inline integration-tests/swiss-knife/app/src/main/blueprint/blueprint.conf:19]:
```suggestion
      producers = [spark-process.out, akka-process.out, flink-process.out]
```

File: core/project/Dependencies.scala [hunk 9]
```diff
@@ -2,26 +2,28 @@ import sbt._
 
 // format: OFF
 object Version {
+
   val Akka         = "2.6.5"
   val AkkaHttp     = "10.1.11"
   val AkkaMgmt     = "1.0.6"
   val AlpakkaKafka = "2.0.3"
   val Scala        = "2.12.9"
```
Comment (yuchaoran2011) [inline core/project/Dependencies.scala:10]: We can make this 2.12.11

File: integration-tests/swiss-knife/build.sbt [hunk 10]
```diff
@@ -2,37 +2,66 @@ import sbt._
...
+  .dependsOn(datamodel)
+
+lazy val flink = (project in file("./flink"))
+  .enablePlugins(CloudflowFlinkPlugin)
+  .settings(commonSettings)
+  .dependsOn(datamodel)
+
+lazy val commonSettings = Seq(
+  headerLicense := Some(HeaderLicense.ALv2("(C) 2016-2020", "Lightbend Inc. <https://www.lightbend.com>")),
+  scalaVersion := "2.12.10",
```
Comment (yuchaoran2011) [inline integration-tests/swiss-knife/build.sbt:54]: Same here, 2.12.11

File: integration-tests/swiss-knife/app/src/main/blueprint/blueprint.conf [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+
+  topics {
+    source-data-pipe {
+      producers = [ingress.out]
+      consumers = [egress.in, akka-process.in, spark-process.in, flink-process.in, spark-output.in]
+    }
+
+    process-pipe {
+      producers = [spark-process.out, akka-process.out, flink-process.out]
+      consumers = [ egress.in ]
```
Comment (yuchaoran2011) [inline integration-tests/swiss-knife/app/src/main/blueprint/blueprint.conf:20]:
```suggestion
      consumers = [egress.in]
```

File: integration-tests/swiss-knife/app/src/main/resources/local.conf [hunk 12]
```diff
@@ -0,0 +1 @@
+#configuration settings used in the local running mode go here.
```
Comment (yuchaoran2011) [inline integration-tests/swiss-knife/app/src/main/resources/local.conf:1]:
```suggestion
#configuration settings used in the local running mode go here.

```

File: core/project/Dependencies.scala (hunk 9 above)
Comment (maasg) [reply core/project/Dependencies.scala:10]: Good idea. Let me upgrade in a separate PR that we can test properly. I don't expect any issues but better to encapsulate that.

Comment (yuchaoran2011) [reply core/project/Dependencies.scala:10]: Sure

**Pull request merged**